File: client/exceptions.py
class ClientError(Exception):
    response = None
    code = None
    message = ""

    def __init__(self, *args, **kwargs):
        args = list(args)
        if len(args) > 0:
            self.message = str(args.pop(0))
        for key in list(kwargs.keys()):
            setattr(self, key, kwargs.pop(key))
        super().__init__(self.message, *args, **kwargs)
        if self.response:
            self.code = self.response.status_code


class GenericRequestError(ClientError):
    """Sorry, there was a problem with your request"""


class ClientGraphqlError(ClientError):
    """Raised due to graphql issues"""


class ClientJSONDecodeError(ClientError):
    """Raised due to json decoding issues"""


class ClientConnectionError(ClientError):
    """Raised due to network connectivity-related issues"""


class ClientBadRequestError(ClientError):
    """Raised due to a HTTP 400 response"""


class ClientForbiddenError(ClientError):
    """Raised due to a HTTP 403 response"""


class ClientNotFoundError(ClientError):
    """Raised due to a HTTP 404 response"""


class ClientThrottledError(ClientError):
    """Raised due to a HTTP 429 response"""


class ClientRequestTimeout(ClientError):
    """Raised due to a HTTP 408 response"""


class ClientIncompleteReadError(ClientError):
    """Raised due to incomplete read HTTP response"""


class ClientLoginRequired(ClientError):
    """Instagram redirect to https://www.instagram.com/accounts/login/
    """


class ReloginAttemptExceeded(ClientError):
    pass


class PrivateError(ClientError):
    """For Private API and last_json logic
    """


class FeedbackRequired(PrivateError):
    pass


class ChallengeError(PrivateError):
    pass


class ChallengeRedirection(ChallengeError):
    pass


class ChallengeRequired(ChallengeError):
    pass


class SelectContactPointRecoveryForm(ChallengeError):
    pass


class RecaptchaChallengeForm(ChallengeError):
    pass


class SubmitPhoneNumberForm(ChallengeError):
    pass


class LoginRequired(PrivateError):
    """Instagram request relogin
    Example:
    {'message': 'login_required',
    'response': <Response [403]>,
    'error_title': "You've Been Logged Out",
    'error_body': 'Please log back in.',
    'logout_reason': 8,
    'status': 'fail'}
    """


class SentryBlock(PrivateError):
    pass


class RateLimitError(PrivateError):
    pass


class BadPassword(PrivateError):
    pass


class UnknownError(PrivateError):
    pass


class MediaError(PrivateError):
    pass


class MediaNotFound(MediaError):
    pass


class UserError(PrivateError):
    pass


class UserNotFound(UserError):
    pass


class CollectionError(PrivateError):
    pass


class CollectionNotFound(CollectionError):
    pass

File: src/exceptions.ts
export class ClientError extends Error {
  response?: any;
  code?: number;

  constructor(message?: string, ...args: any[]) {
    super(message, ...args);
    this.name = this.constructor.name;
  }
}

export class ClientUnknownError extends ClientError {}
export class WrongCursorError extends ClientError {
  message = "You specified a non-existent cursor";
}
export class ClientStatusFail extends ClientError {}
export class ClientErrorWithTitle extends ClientError {}
export class ResetPasswordError extends ClientError {}
export class GenericRequestError extends ClientError {}
export class ClientGraphqlError extends ClientError {}
export class ClientJSONDecodeError extends ClientError {}
export class ClientConnectionError extends ClientError {}
export class ClientBadRequestError extends ClientError {}
export class ClientUnauthorizedError extends ClientError {}
export class ClientForbiddenError extends ClientError {}
export class ClientNotFoundError extends ClientError {}
export class ClientThrottledError extends ClientError {}
export class ClientRequestTimeout extends ClientError {}
export class ClientIncompleteReadError extends ClientError {}
export class ClientLoginRequired extends ClientError {}
export class ReloginAttemptExceeded extends ClientError {}
export class PrivateError extends ClientError {}
export class NotFoundError extends PrivateError {
  reason = "Not found";
}

export class HashtagError extends PrivateError {}
export class HashtagNotFound extends NotFoundError {}

export class UserError extends PrivateError {}
export class UserNotFound extends NotFoundError {}


export class FeedbackRequired extends PrivateError {}
export class ChallengeError extends PrivateError {}
export class ChallengeRedirection extends ChallengeError {}
export class ChallengeRequired extends ChallengeError {}
export class ChallengeSelfieCaptcha extends ChallengeError {}
export class ChallengeUnknownStep extends ChallengeError {}
export class SelectContactPointRecoveryForm extends ChallengeError {}
export class RecaptchaChallengeForm extends ChallengeError {}
export class SubmitPhoneNumberForm extends ChallengeError {}
export class LegacyForceSetNewPasswordForm extends ChallengeError {}
export class LoginRequired extends PrivateError {}
export class SentryBlock extends PrivateError {}
export class RateLimitError extends PrivateError {}
export class ProxyAddressIsBlocked extends PrivateError {}
export class BadPassword extends PrivateError {}
export class BadCredentials extends PrivateError {}
export class PleaseWaitFewMinutes extends PrivateError {}
export class UnknownError extends PrivateError {}
export class TwoFactorRequired extends PrivateError {}
export class PrivateAccount extends PrivateError {}
export class InvalidTargetUser extends PrivateError {}
export class InvalidMediaId extends PrivateError {}
export class MediaUnavailable extends PrivateError {}
export class ValidationError extends Error {}
export class EmailInvalidError extends ClientError {}
export class EmailNotAvailableError extends ClientError {}
export class EmailVerificationSendError extends ClientError {}
export class AgeEligibilityError extends ClientError {}
export class CaptchaChallengeRequired extends ClientError {
    challenge_details: any;
    constructor(message = "Captcha challenge required, but no solver configured or available.", challenge_details = {}, ...args: any[]) {
        super(message, ...args);
        this.challenge_details = challenge_details;
    }
}
